docs/design/garbage-collection.md

```diff
@@ -145,3 +146,68 @@ see: [Bloom filter math](http://pages.cs.wisc.edu/~cao/papers/summary-cache/node
     - Storage node should be able to receive a Delete request from a Satellite
 - Storage node should use the filter from the Delete request to decide which pieces to delete, then delete them
 - Eventually, this service should iterate over a db snapshot instead of being integrated with the data repair checker
+
+## Bloom filters benchmark
+
+Three Bloom filter implementations are considered:
+- **BF1**: Zeebo's bloom filters (github.com/zeebo/sbloom)
+- **BF2**: Willf's Bloom filters (github.com/zeebo/sbloom)
+- **BF3**: Steakknife's Bloom filters (github.com/golang/leveldb/bloom)
+- **BF4**: Custom bloom filter
+
+### Zeebo's bloom filters
+-Parameters:
+    - **k**: The Bloom filter will be built such that the probability of a false positive is less than (1/2)**k
+    - **h**: hash functions
+- Serialization available
+- hash functions are to be given as a parameter to the constructor
+### Willf's bloom filters
+- Parameters:
+    - **m**: max size in bits
+    - **k**: number of hash functions
+- hash functions not configurable
+
+### Steakknife's bloom filters
+- Parameters:
+    - **maxElements**: max number of elements in the set
+    - **p**: probability of false positive
+- Serialization available
+- murmur3 hash function
+
+### Custom bloom filter
+- Parameters:
+    - **maxElements**: max number of elements in the set
```

Comment:
Does this mean that when a bloom filter is created, the caller needs to know that maximum possible number of element that will be added to the filter?

If this is the case then I am not sure if this will work well with the satellite. Initially, we will make the data repair checker to create the bloom filters as part of its usual work. I am not sure if it knows how many pieces are stored on each storage node.

Is it possible to construct the custom bloom filter in a way that the `maxElements` is not known?

Reply:
I thought we would know the bloom filter size beforehand. The custom bloom filter does not handle that (because then we cannot compute the optimal values for k and m), I guess in this case using Zeebo's implementation would make more sense.

Reply:
We can keep the last number around and then use 2x of that number for the next filter.

Reply:
What happens if we add more than `maxElements` to the filter. Will it become corrupted?

Reply:
No, no corruption, just the false positive rate will be terrible.

Reply:
So, let's say, if we initialize the bloom filter with 1000 max elements, but add 2000, it would still be worth sending it to the storage node. It will be just less effective, right?

Reply:
It will be less effective and at some point it would make sense to just drop it for that iteration.

Reply:
e.g. if you have FP=2%, and 1M items, then inserting 2M will give you FP=14%, with 3M the FP=33%

Reply:
OK. Cool. So we can actually just use the last known piece count for this node from the last iteration. It will just result in higher FP if it happens that this node stores more pieces. No need to waste memory with 2x last known piece count.

Reply:
Yeah, that would work as well.